pyvisgen/simulation/visibility.py

```diff
@@ -1,222 +1,176 @@
-                lm,
-                baselines,
-                IF,
-                t,
-                src_crd,
-                array_layout,
-                SI,
-                rd,
-                vis_num,
-                corrupted=rc["corrupted"],
-            )
-            int_values.append(val_i)
-            del val_i
-
-        int_values = np.array(int_values)
-        if int_values.dtype != np.complex128:
+        if int_values.numel() == 0:
             continue
-        int_values = np.swapaxes(int_values, 0, 1)
 
-        if noisy:
-            noise = generate_noise(int_values.shape, rc)
+        int_values = torch.swapaxes(int_values, 0, 1)
+
+        if noisy != 0:
+            noise = generate_noise(int_values.shape, obs, noisy)
             int_values += noise
 
-        vis_num = np.arange(int_values.shape[0]) + 1 + vis_num.max()
+        vis_num = torch.arange(int_values.shape[0]) + 1 + vis_num.max()
 
         vis = Visibilities(
-            torch.tensor(int_values[:, :, 0]),
-            torch.zeros(int_values[:, :, 0].shape, dtype=torch.complex128),
-            torch.zeros(int_values[:, :, 0].shape, dtype=torch.complex128),
-            torch.zeros(int_values[:, :, 0].shape, dtype=torch.complex128),
+            int_values[:, :, 0, 0].cpu(),
+            int_values[:, :, 0, 1].cpu(),
+            int_values[:, :, 1, 0].cpu(),
+            int_values[:, :, 1, 1].cpu(),
             vis_num,
-            np.repeat(i + 1, len(vis_num)),
-            np.array([baselines[i].baselineNum() for i in base_valid]),
-            u_valid,
-            v_valid,
-            w_valid,
-            date,
-            _date,
+            bas_p[9].cpu(),
+            bas_p[2].cpu(),
+            bas_p[5].cpu(),
+            bas_p[8].cpu(),
+            bas_p[10].cpu(),
         )
 
         visibilities.add(vis)
         del int_values
     return visibilities
 
 
-def calc_vis(
-    lm, baselines, wave, t, src_crd, array_layout, SI, rd, vis_num, corrupted=True
-):
+def calc_vis(bas, lm, rd, ra, dec, ant_diam, spw_low, spw_high, SI, corrupted=False):
     if corrupted:
-        X1 = scan.direction_independent(
-            lm, baselines, wave, t, src_crd, array_layout, SI, rd
-        )
-        if X1.shape[0] == 1:
-            return -1
-        X2 = scan.direction_independent(
-            lm, baselines, wave, t, src_crd, array_layout, SI, rd
+        int_values = scan.rime(
+            SI, bas, lm, rd, ra, dec, ant_diam, spw_low, spw_high, corrupted=corrupted
         )
     else:
-        X1 = scan.uncorrupted(lm, baselines, wave, t, src_crd, array_layout, SI)
-        if X1.shape[0] == 1:
-            return -1
-        X2 = scan.uncorrupted(lm, baselines, wave, t, src_crd, array_layout, SI)
-
-    int_values = scan.integrate(X1, X2).numpy()
-    del X1, X2, SI
-    int_values = int_values
+        int_values = scan.rime(SI, bas, lm, rd, ra, dec, ant_diam, spw_low, spw_high)
     return int_values
```

Comment:
Is it necessary to have a function for this? Because the query just passes the keyword, which can be true or false. So when the function is called, one could just call `scan.rime(..., corrupted=corrupted)` without the need for the `calc_vis` function.

Reply:
danone